L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc

```diff
@@ -0,0 +1,905 @@
+  geom_ordered_set getOrderedTriggerCellsFromModule(const unsigned) const final;
+
+  geom_set getNeighborsFromTriggerCell(const unsigned) const final;
+
+  unsigned getLinksInModule(const unsigned module_id) const final;
+  unsigned getModuleSize(const unsigned module_id) const final;
+
+  GlobalPoint getTriggerCellPosition(const unsigned) const final;
+  GlobalPoint getModulePosition(const unsigned) const final;
+
+  bool validCell(const unsigned) const final;
+  bool validTriggerCell(const unsigned) const final;
+  bool disconnectedModule(const unsigned) const final;
+  unsigned lastTriggerLayer() const final { return last_trigger_layer_; }
+  unsigned triggerLayer(const unsigned) const final;
+
+private:
+  // HSc trigger cell grouping
+  unsigned hSc_triggercell_size_ = 2;
+  unsigned hSc_module_size_ = 12;  // in TC units (144 TC / panel = 36 e-links)
+  unsigned hSc_links_per_module_ = 1;
+  unsigned hSc_wafers_per_module_ = 3;
+
+  unsigned sector0_mask_ = 0x7f;  // 7 bits to encode module number in 60deg sector
+
+  edm::FileInPath l1tModulesMapping_;
+  edm::FileInPath l1tLinksMapping_;
+  edm::FileInPath jsonMappingFile_;
+
+  // module related maps
+  std::unordered_map<unsigned, unsigned> wafer_to_module_;
+  std::unordered_multimap<unsigned, unsigned> module_to_wafers_;
+  std::unordered_map<unsigned, unsigned> links_per_module_;
+  mutable tbb::concurrent_unordered_set<unsigned> cache_missing_wafers_;
+
+  json mapping_config_;
+
+  // Disconnected modules and layers
+  std::unordered_set<unsigned> disconnected_modules_;
+  std::unordered_set<unsigned> disconnected_layers_;
+  std::vector<unsigned> trigger_layers_;
+  std::vector<unsigned> trigger_nose_layers_;
+  unsigned last_trigger_layer_ = 0;
+
+  // layer offsets
+  unsigned heOffset_ = 0;
+  unsigned noseLayers_ = 0;
+  unsigned totalLayers_ = 0;
+
+  void loadJsonMappingFile();
+  void fillMaps();
+  bool validCellId(unsigned det, unsigned cell_id) const;
+  bool validTriggerCellFromCells(const unsigned) const;
+
+  int detIdWaferType(unsigned det, unsigned layer, short waferU, short waferV) const;
+  unsigned packWaferId(int waferU, int waferV) const;
+  unsigned packLayerWaferId(unsigned layer, int waferU, int waferV) const;
+  unsigned packLayerModuleId(unsigned layer, unsigned wafer) const;
+  void unpackWaferId(unsigned wafer, int& waferU, int& waferV) const;
+
+  unsigned layerWithOffset(unsigned) const;
+
+  std::vector<unsigned> getStage1FpgasFromStage2Fpga(const unsigned stage2_id) const;
+  std::vector<unsigned> getLbgbtsFromStage1Fpga(const unsigned stage1_id) const;
+  std::vector<unsigned> getStage2FpgasFromStage1Fpga(const unsigned stage1_id) const;
+  unsigned getStage1FpgaFromLpgbt(const unsigned lpgbt_id) const;
+  std::vector<std::tuple<bool, int, int, unsigned> > getModulesFromLpgbt(const unsigned lpgbt_id) const;
+  std::vector<unsigned> getLpgbtsFromModule(const bool isSilicon, const int module_u, const int module_v, const unsigned layer) const;
```

Comment:
Module ids (integer) should be defined instead of accessing the mappings with tuples. This is to reduce as much as possible CPU time. See also my other comments below on the access functions.